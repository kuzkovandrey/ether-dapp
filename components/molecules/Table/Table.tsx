'use client';

import { Flex, Heading, Table as RdxTable } from '@radix-ui/themes';
import { ReactNode, useMemo } from 'react';

import { Pagination } from '@/components/atoms';

export type TableColumn<D> = {
  key: keyof D extends string ? keyof D : never;
  title: ReactNode;
  render: (d: D) => ReactNode;
};

export type TableColumns<D> = Array<TableColumn<D>>;

export const NONE_VALUE = '--';

export type TableProps<D> = {
  heading: string;
  pagination: {
    page: number;
    pages: number;
    rowsPerPage: number;
    onClickNext: () => void;
    onClickPrev: () => void;
  };
  data: Array<D>;
  columns: TableColumns<D>;
  rowsPerPage?: number;
};

function Table<D>({ heading, data, pagination, columns }: TableProps<D>) {
  const displayData = useMemo(() => {
    const startIndex = (pagination.page - 1) * pagination.rowsPerPage;

    return data.slice(startIndex, startIndex + pagination.rowsPerPage);
  }, [data, pagination]);

  return (
    <Flex direction="column" gap="2">
      <Heading size="4">{heading}</Heading>
      <RdxTable.Root variant="surface">
        <RdxTable.Header>
          <RdxTable.Row>
            {Object.values(columns).map(({ key, title }) => (
              <RdxTable.ColumnHeaderCell key={`${key}-hCell`}>{title}</RdxTable.ColumnHeaderCell>
            ))}
          </RdxTable.Row>
        </RdxTable.Header>

        <RdxTable.Body>
          {Array.from(displayData).map((d, index) => (
            <RdxTable.Row key={`${index}-bodyRow`}>
              {columns.map((column, index) => (
                <RdxTable.Cell key={`${index}-cell`}>{column.render(d)}</RdxTable.Cell>
              ))}
            </RdxTable.Row>
          ))}
        </RdxTable.Body>
      </RdxTable.Root>

      {Boolean(data.length) && (
        <Pagination
          page={pagination.page}
          pages={pagination.pages}
          onNext={pagination.onClickNext}
          onPrev={pagination.onClickPrev}
        />
      )}
    </Flex>
  );
}

export default Table;
