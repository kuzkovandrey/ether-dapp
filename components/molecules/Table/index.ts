export { NONE_VALUE, type PaginationProps, default as Table, type TableColumns } from './Table';
