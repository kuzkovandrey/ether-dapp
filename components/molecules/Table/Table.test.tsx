/**
 * @jest-environment jsdom
 */

import { cleanup, render, screen } from '@testing-library/react';

import '@testing-library/jest-dom';
import Table, { TableProps } from './Table';

afterEach(cleanup);

type TestData = {
  field1: string;
  field2: number;
  field3: string;
};

const mockData: TestData[] = [
  { field1: 'field11', field2: 21, field3: '31' },
  { field1: 'field12', field2: 22, field3: '32' },
  { field1: 'field13', field2: 23, field3: '33' },
];

describe('Table component', () => {
  let initialProps = {} as TableProps<TestData>;

  beforeEach(() => {
    initialProps = {
      heading: 'Heading text',
      pagination: {
        page: 1,
        pages: 1,
        rowsPerPage: 5,
        onClickNext: jest.fn(),
        onClickPrev: jest.fn(),
      },
      isLoading: false,
      data: [],
      columns: [
        {
          key: 'field1',
          title: 'field title',
          render: ({ field1 }) => <>{field1}</>,
        },
        {
          key: 'field2',
          title: <div>field2 title</div>,
          render: ({ field2 }) => <div>{field2}</div>,
        },
        {
          key: 'field3',
          title: <p>field3 title</p>,
          render: ({ field3 }) => <p>{field3}</p>,
        },
      ],
    };
  });

  test('# empty component should be rendered on the screen', () => {
    render(<Table {...initialProps} />);
    expect(screen.getByTestId('table')).toBeInTheDocument();
    expect(screen.getByTestId('table-heading')).toHaveTextContent(initialProps.heading);
    expect(screen.getByTestId('table-nothing-found')).toHaveTextContent(/Nothing found/i);
    expect(screen.queryByTestId('table-loader')).not.toBeInTheDocument();
    expect(screen.queryByTestId('table-pagination')).not.toBeInTheDocument();
    expect(screen.queryAllByTestId('table-header-cell').length).toBe(0);
    expect(screen.queryAllByTestId('table-row').length).toBe(0);
  });

  test('# component must have header cells with text content', () => {
    initialProps.data = mockData;
    render(<Table {...initialProps} />);

    const cells = screen.getAllByTestId('table-header-cell');

    expect(cells.length).toBe(initialProps.columns.length);
    cells.forEach((cell) => expect(cell).toMatchSnapshot());
  });

  test('# component must have loader', () => {
    const { rerender } = render(<Table {...initialProps} />);

    expect(screen.queryByTestId('table-loader')).not.toBeInTheDocument();

    initialProps.isLoading = true;
    rerender(<Table {...initialProps} />);
    expect(screen.getByTestId('table-loader')).toBeInTheDocument();
  });

  test('# component must have data content', () => {
    initialProps.data = mockData;
    const { rerender } = render(<Table {...initialProps} />);

    expect(screen.getAllByTestId('table-row').length).toBe(mockData.length);
    expect(screen.queryByTestId('table-nothing-found')).not.toBeInTheDocument();

    initialProps.data = [...mockData, ...mockData];
    rerender(<Table {...initialProps} />);

    const rows = screen.getAllByTestId('table-row');

    expect(rows.length).toBe(initialProps.pagination.rowsPerPage);
    rows.forEach((row) => expect(row).toMatchSnapshot());
  });
});
