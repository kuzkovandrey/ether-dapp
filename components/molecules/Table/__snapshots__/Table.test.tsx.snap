// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table component # component must have data content 1`] = `
<tr
  class="rt-TableRow"
  data-testid="table-row"
>
  <td
    class="rt-TableCell"
  >
    field11
  </td>
  <td
    class="rt-TableCell"
  >
    <div>
      21
    </div>
  </td>
  <td
    class="rt-TableCell"
  >
    <p>
      31
    </p>
  </td>
</tr>
`;

exports[`Table component # component must have data content 2`] = `
<tr
  class="rt-TableRow"
  data-testid="table-row"
>
  <td
    class="rt-TableCell"
  >
    field12
  </td>
  <td
    class="rt-TableCell"
  >
    <div>
      22
    </div>
  </td>
  <td
    class="rt-TableCell"
  >
    <p>
      32
    </p>
  </td>
</tr>
`;

exports[`Table component # component must have data content 3`] = `
<tr
  class="rt-TableRow"
  data-testid="table-row"
>
  <td
    class="rt-TableCell"
  >
    field13
  </td>
  <td
    class="rt-TableCell"
  >
    <div>
      23
    </div>
  </td>
  <td
    class="rt-TableCell"
  >
    <p>
      33
    </p>
  </td>
</tr>
`;

exports[`Table component # component must have data content 4`] = `
<tr
  class="rt-TableRow"
  data-testid="table-row"
>
  <td
    class="rt-TableCell"
  >
    field11
  </td>
  <td
    class="rt-TableCell"
  >
    <div>
      21
    </div>
  </td>
  <td
    class="rt-TableCell"
  >
    <p>
      31
    </p>
  </td>
</tr>
`;

exports[`Table component # component must have data content 5`] = `
<tr
  class="rt-TableRow"
  data-testid="table-row"
>
  <td
    class="rt-TableCell"
  >
    field12
  </td>
  <td
    class="rt-TableCell"
  >
    <div>
      22
    </div>
  </td>
  <td
    class="rt-TableCell"
  >
    <p>
      32
    </p>
  </td>
</tr>
`;

exports[`Table component # component must have header cells with text content 1`] = `
<th
  class="rt-TableCell rt-TableColumnHeaderCell"
  data-testid="table-header-cell"
  scope="col"
>
  field title
</th>
`;

exports[`Table component # component must have header cells with text content 2`] = `
<th
  class="rt-TableCell rt-TableColumnHeaderCell"
  data-testid="table-header-cell"
  scope="col"
>
  <div>
    field2 title
  </div>
</th>
`;

exports[`Table component # component must have header cells with text content 3`] = `
<th
  class="rt-TableCell rt-TableColumnHeaderCell"
  data-testid="table-header-cell"
  scope="col"
>
  <p>
    field3 title
  </p>
</th>
`;
